java/clients/src/main/java/sleeper/clients/QueryWebSocketClient.java

```diff
@@ -73,30 +80,39 @@ public class QueryWebSocketClient {
         this.out = out;
     }
 
-    public void submitQuery(Query query) {
+    public CompletableFuture<List<String>> submitQuery(Query query) throws InterruptedException {
         try {
-            Instant startTime = Instant.now();
-            client.startQuery(query);
-            while (!client.hasQueryFinished()) {
-                Thread.sleep(500);
-            }
-            LoggedDuration duration = LoggedDuration.withFullOutput(startTime, Instant.now());
-            long recordsReturned = client.getTotalRecordsReturned();
-            out.println("Query took " + duration + " to return " + recordsReturned + " records");
+            startTime = Instant.now();
+            return client.startQueryFuture(query)
+                    .whenComplete((records, exception) -> {
+                        try {
+                            client.closeBlocking();
+                        } catch (InterruptedException e) {
+                        }
```

Comment:
An InterruptedException should cause the thread to terminate immediately. This could rethrow and re-interrupt the thread.

Reply:
Thanks, i'm rethrowing this exception out of the method

Reply:
When we rethrow a wrapped InterruptedException we also need to re-interrupt the thread with `Thread.currentThread().interrupt()`.

If the InterruptedException is wrapped as a RuntimeException then any code that catches it can't know that it needs to terminate immediately, unless we also re-interrupt the thread.

Reply:
Ah ok thanks, I've included that manual interruption in this block